test/test_util.jl

```diff
@@ -0,0 +1,188 @@
+using LinearAlgebra: AbstractTriangular
+
+#####
+##### We need to be able to convert everything into into a vector for use with FDM.jl
+#####
+
+# Transform `x` into a vector, and return a closure which inverts the transformation.
+to_vec(x::Nothing) = (x, nothing)
+to_vec(x::Real) = ([x], x->x[1])
+
+# Arrays.
+to_vec(x::Vector{<:Real}) = (x, identity)
+to_vec(x::Array) = vec(x), x_vec->reshape(x_vec, size(x))
+
+# AbstractArrays.
+function to_vec(x::T) where {T<:AbstractTriangular}
+    x_vec, back = to_vec(Matrix(x))
+    return x_vec, x_vec->T(reshape(back(x_vec), size(x)))
+end
+to_vec(x::Symmetric) = vec(Matrix(x)), x_vec->Symmetric(reshape(x_vec, size(x)))
+to_vec(X::Diagonal) = vec(Matrix(X)), x_vec->Diagonal(reshape(x_vec, size(X)...))
+
+# Non-array data structures.
+function to_vec(x::Tuple)
+    x_vecs, x_backs = zip(map(to_vec, x)...)
+    sz = cumsum([map(length, x_vecs)...])
+    return vcat(x_vecs...), v->([x_backs[n](v[sz[n]-length(x[n])+1:sz[n]]) for n in 1:length(x)]...,)
+end
+
+function FDM.jvp(fdm, f, x, ẋ)
```

Comment:
Seems like this should be defined in FDM instead, especially since here we're extending a function from FDM without providing types defined in this package (actually no types at all), which is Type Piracy™.